sound/soc/sof/sof-client-probes-ipc4.c

```diff
@@ -0,0 +1,276 @@
+	msg.extension |= SOF_IPC4_MOD_EXT_CORE_ID(0);
+
+	msg.data_size = 0;
+	msg.data_ptr = NULL;
+
+	ret = sof_client_ipc_tx_message(cdev, &msg, NULL, 0);
+	if (ret)
+		dev_err(dev, "%s() failed with: %d", __func__, ret);
+
+	return ret;
+}
+
+static int ipc4_probes_points_info(struct sof_client_dev *cdev,
+				   struct sof_probe_point_desc **desc,
+				   size_t *num_desc)
+{
+	/* TODO: Firmware side implementation needed first */
+	*desc = NULL;
+	*num_desc = 0;
+	return 0;
+}
+
+static int ipc4_probes_points_add(struct sof_client_dev *cdev,
+				  struct sof_probe_point_desc *desc,
+				  size_t num_desc)
+{
+	struct device *dev = &cdev->auxdev.dev;
+	struct sof_man4_module *mentry =
+		sof_ipc4_probe_get_module_info(cdev);
+	struct sof_ipc4_msg msg;
+	struct sof_ipc4_probe_point *points;
+	int i, ret;
+
+	if (!mentry)
+		return -EINVAL;
+
+	/* The sof_probe_point_desc and sof_ipc4_probe_point structs are
+	 * of same size and even the integers are same in same order, and
+	 * similar meaning, but since there is no performance issue wrote
+	 * the conversion explicitly open for future development.
+	 */
+	points = kzalloc(num_desc * sizeof(*points), GFP_KERNEL);
+	if (!points)
+		return -ENOMEM;
+
+	for (i=0; i < num_desc; i++) {
+		points[i].buffer_id.full_id = desc[i].buffer_id;
+		points[i].purpose = desc[i].purpose;
+		points[i].stream_tag = desc[i].stream_tag;
+	}
+
+	msg.primary = mentry->id;
+	msg.primary |= SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_MOD_LARGE_CONFIG_SET);
+	msg.primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
+	msg.primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_MODULE_MSG);
+
+	msg.extension = SOF_IPC4_MOD_EXT_MSG_PARAM_ID(SOF_IPC4_PROBE_POINTS);
+
+	msg.data_size = sizeof(*points) * num_desc;
+	msg.data_ptr = points;
+
+	ret = sof_client_ipc_set_get_data(cdev, &msg, true);
+	if (ret)
+		dev_err(dev, "%s() failed with: %d", __func__, ret);
+
+	kfree(points);
+
+	return ret;
+}
+
+static int ipc4_probes_points_remove(struct sof_client_dev *cdev,
+				     unsigned int *buffer_id, size_t num_buffer_id)
+{
+	struct device *dev = &cdev->auxdev.dev;
+	struct sof_man4_module *mentry =
+		sof_ipc4_probe_get_module_info(cdev);
+	struct sof_ipc4_msg msg;
+	union sof_ipc4_probe_point_id *ids;
+	int i, ret;
+
+	if (!mentry)
+		return -EINVAL;
+
+	ids = kzalloc(num_buffer_id * sizeof(*ids), GFP_KERNEL);
+	if (!ids)
+		return -ENOMEM;
+
+	for (i=0; i < num_buffer_id; i++)
+		ids[i].full_id = buffer_id[i];
+
+	msg.primary = mentry->id;
+	msg.primary |= SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_MOD_LARGE_CONFIG_SET);
+	msg.primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
+	msg.primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_MODULE_MSG);
+
+	msg.extension =
+		SOF_IPC4_MOD_EXT_MSG_PARAM_ID(SOF_IPC4_PROBE_POINTS_DISCONNECT);
+
+	msg.data_size = num_buffer_id * sizeof(union sof_ipc4_probe_point_id);
+	msg.data_ptr = ids;
+
+	ret = sof_client_ipc_set_get_data(cdev, &msg, true);
+	if (ret)
+		dev_err(dev, "%s() failed with: %d", __func__, ret);
```

Comment:
Same here, I think we are going to be aware of the error codes from other layers.